### DISCUSSION TITLE
Implementation design for ancestral lines

### DISCUSSION THREAD

**Discussion by ilyaZar** [rcppsmc/rcppsmc#55] [Summer of Code]
So this topic is meant to start the discussion about exactly how we might want to implement / deal with ancestral lines.

My thoughts on this are around two things: (1.) what to store and (2.) where to store (1.).

1.  _**What to store:**_
A. with N being the number of particles and lIterates the number of SMC iterations, a naïve storage facility would copy the ancestor indices (e.g. into an `arma::umat (N, lIterates)` ) and plain particle values (e.g. into `std::array<std::vector<Space, N>, lIterates>`)
B.	another way of providing storage would be to compute ancestral lines at each `T=0,…,lIterates` and directly store the result (e.g. into `std::array<std::vector<Space, N>, lIterates>`)
C.	the different quantities associated with Alg2 and Alg3 of the “Path storage in the particle filter” paper 
2. _**Where to store:**_
A.	Redesign (or make a derived) class of the current `historyelement`-class with additional data fields (whatever has been decided upon under 1.) to store ancestral lines there. Along with this derived class, say `historyelementAL`, we would also require a switch at the end of the `sampler<Space, Param>::Initiliaze()` and `sampler<Space, Param>::IterateESS()` member functions of the type:
    ```
    if(htHistoryMode = HistoryType::NONE) {
    // ...behaviour as before...
    } else if (htHistoryMode = HistoryType::AL) {
    // ...behaviour as before but only a difference to `historyelementAL.Set()` member function that operates on the derived 
   // historyelementAL-class additionally setting private data fields `historyelementAL` (whatever has been decided upon under 1)
   // accordingly
   }
    ```
    Each `historyelement` (and similarly also the `historyelementAL`) would be stored in  
    `std::vector<historyelementAL<Space> > History` automatically.
B.	redesign (or make a derived) class of the `population`-class and store (whatever has been decided on under 1.) in (private) data fields there. Currently the main private data fields are `population.value` (storing the resampled particle values) and `population.logweight` (storing the log-weights), thus adjustments are needed there (storing whatever has been decided upon under 1.)..

In 2.A we would need interaction from the resampling facilities (which need adjustments anyways) to the `historyelement`-class while leaving the `population`-class as is. 

In 2.B the `historyelement`-class and associated facilities could be left as is: the behaviour of the `historyelement`-class stores the whole population class atm (and thus would store the ancestors lines, along with `population.value` and `population.logweight`), hence only interactions between `population`-class and resampling facilities (which need adjustments anyways) require changes.

I am not sure which route to take here, 2.A or 2.B.

Somehow it feels to me that leaving the `population`-class as is (i.e. going for choice 2.A) is safer than tweaking the `population`-class (2.B), since there is more dependence of other code parts on  the `population`-class (so more possibilities to accidently break sth.). Contrary, `historyelement` are only used at the end of `Initialise()` and `IterateESS()` for storage purposes and thus only these parts can break. Also, making ancestral lines part of the `population`-class might make the current implementation slightly less efficient if storage of ancestral lines is not required. However, making ancestral lines an inherent data field of the `population`-class (going for route 2.B)  might be easier in terms of implementation (one could alter these fields at the very end of the `Resampling` function and probably that's almost it).

Hopefully this is understandable and helpful somehow as a discussion start.

**Comment by LeahPrice** [comment]
Thanks @ilyaZar for this! I'll think more about it over the next day and I should be able to respond with my thoughts before your end of day on Monday.

**Comment by adamjohansen** [comment]
Thanks @ilyaZar this all sounds very sensible and I imagine most of these approaches could be made to work.

In terms of what to store, my impression is that C is rather harder to do than A and B and although it has some advantages I'd favour keeping it at something we could aim to implement once we have a more basic implementation working. It's more complex, but brings with it a lower space complexity than the more naive strategy. I personally think A would be a fairly natural and fairly straightforward way to proceed. I'm not 100% sure of the details you have in mind with B but it looks as though it would require more manipulation as the algorithm progresses to keep it correct. Even if one has C there is value in having A available from the point of view of allowing things like forward filtering backward simulation or backward sampling variants of conditional SMC. So I'd suggest A with C as something that it would be nice to have as a more space-efficient alternative if we have time and energy to spare in the future.

Where to store things is perhaps more delicate. Instinctively, my immediate reaction was to prefer your option A. (The ancestral lines fairly fundamentally describe the history of the system and in a sense provide an alternative to storing the history of each particle explicitly by storing either the entire history of the particle system and a genealogical forest showing which element at any given generation is the parent of each child or the more efficient tree-based version of the same philosophy.) It would take a bit of rethinking the way the mechanics of the history work, but I think that's a good thing: at present doing smoothing or anything like it really requires a lot of extra effort from the user perspective (essentially you end up working on the historical-process state space or doing something more involved) and we could fix that by properly storing information about ancestral lines.

From a higher-level perspective, one thing we want to have in mind is how easy it will be to implement a conditional version of the SMC algorithm (probably via a derived class which overloads the sampling and resampling operations) using whatever method we have to store ancestral lines here. I can't immediately think of anything in doing this which would strongly favour any of these alternatives, but I'm saying it now just in case anyone else can.

So, what does everyone else thing? (It hopefully goes without saying, that  other views are very welcome... I haven't had time to think all that hard about this and, if I'm overlooking things, then now would be a good time to find out.)

**Comment by LeahPrice** [comment]
Thanks Ilya and Adam, I feel like this is really productive discussion.

I'm leaning towards 1A and 2A at the moment, and I agree with Adam about holding off on 1C until later.

I may have misinterpreted 1A, but I see that you mentioned storing an `arma::umat (N, lIterates)` and a `std::array<std::vector<Space, N>, lIterates>` object. Fortunately, line 126 of sampler.h already sets us up for the storage of our desired quantities over the `lIterates` with `std::vector<historyelement<Space> > History;`. 

I wonder if the following approach based very much on 1A and 2A might work:
1. In history.h, add a derived class of the `historyelement` class so that we can also store the vector of most recent ancestors. The redesign option that you also suggested might work, but it's less clear to me how we'd handle this when we don't want to store the ancestral lines (what do people think about this)?
2. In sampler.h, have a new `HistoryType` option, include the most recent ancestor indices in the history update and add getters for this new aspect of the history. Of course we also need to get the resampling function to produce the indices, as you've both already mentioned.

What do people think? @adamjohansen, does this go against what you're suggesting in that you'd still be working with the history to do smoothing and other similar things?

**Reply by adamjohansen** [reply]
That's a very good point and I entirely agree.
I'd overlooked the fact that the existing history structure wasn't being used in the described approach and agree it would make more sense to include it within it (that's actually what I had in mind when suggesting a variant of the historyelement class to deal with ancestries a while back).

**Comment by ilyaZar** [comment]
Hi all,

thanks for the insightful an valuable remarks and ideas! (and sorry for the late reply).

Thanks @adamjohansen, I indeed believe you are right when starting with 1.A and leaving 1.C for later. Regarding that we want a conditional SMC at the end of the day: yes, absolutely, (but neither could I think of anything that would obstruct the construction of a derived sampler class in either of the cases 1.A to 1.C).

What @LeahPrice mapped out as a more concrete plan towards implementation with points 1. and 2. is a really good way to start, I believe, thanks for this! (and yes! now that I read it again `std::array<std::vector<Space, N>, lIterates>` does not make any sense to me either :) and actually I saw how the l.126 implementation works and was referencing it as the last bit of 2.A so god knows what my idea was with that...)

I guess I am good to go now, I'll start coding and testing a bit around, and fire up a new PR this week, so we have something more concrete to discuss (I will keep you posted with more details, and in any case, we are having the call tomorrow where we can discuss more!)

Or do we have anything else to discuss? I think we all pretty much settled on 1.A and 2.A.

**Reply by adamjohansen** [reply]
It sounds like you've got a good plan to me; perhaps see how things look when you make a start and we can discuss more when we see what emerges.

**Reply by adamjohansen** [reply]
How are things going? Is the practice proving less simple than the theory suggested?

**Reply by ilyaZar** [reply]
Sorry, for being a bit of an underground worker in the last two weeks. I was reading into several technical programming things. 

Regarding our current task, I will shortly come up with what I managed so far. 

I recall very well the remark of programming being more engineering (trial and error-ish) than doing math in front of a blackboard. Unfortunately, it takes me a while to adjust to this, as you can see.

**Reply by eddelbuettel** [reply]
Let us know how we can help.  As the saying goes, "there are no wrong or silly questions".

**Reply by ilyaZar** [reply]
### I. So to summarize:

1. We have a class template  `template <class Space>class historyelement{/*...*/};`
2. I have implemented a derived class template `template <class Space> class historyelementAL: public historyelement<Space> {/*...*/};` inheriting all members from its base (additionally, it has a storage vector for indices and getter/setter member functions for these indices).
3. The aim is to adjust `std::vector<historyelement<Space> > History;` in a polymorphic way such that it can store instantiated objects of `historyelement<Space>` (if HistoryType::RAM) or  instances of the derived `historyelementAL<Space>` (if HistoryType::AL).

### II. Problem:

The current problem is that of object slicing i.e. I am currently (when setting `HistoryType::AL`) storing objects of the derived class in an vector which is supposed to store objects of the base class. 

Thus the derived class specific members of the object being stored get sliced off, and consequently, the object stored in the vector just acts as object of the base class...

### III. Possible solution:

Now a possible solution is (see e.g. here: https://stackoverflow.com/questions/8777724/store-derived-class-objects-in-base-class-variables) to store pointers `std::vector<historyelement<Space>* > History;`  or, to circumvent possible memory allocation issues, one would probably prefer to use smart pointers as suggested here: https://stackoverflow.com/questions/8706192/which-kind-of-pointer-do-i-use-when. 

And this is where I am, thinking about what to use, and how this might alter the other code parts that  depend on `std::vector<historyelement<Space> > History;` when changing it to (whatever type of) pointer.

### IV. Remarks/current code:

If you fancy something concrete, I can do a draft pull request with preliminary code so you have something visual.

However, I'll have to comment out the members of the derived `historyelementAL` as there are not visible (otherwise there's an error at compile time...) but you can inspect the other changes. 

Obviously, this draft pull request will be **_extremely drafty_**.

As a side remark, I also implemented a backward computation of ancestral lines from indices and particle values, though this facility cannot be used atm because of the above pending issues.

**Reply by LeahPrice** [reply]
Thanks @ilyaZar for the update.

I think I'd find it useful to be able to see the code but I don't think we should create a pull request yet. Perhaps you could create a fork on your own repository and include the link. Would you be okay with doing that?

Based on your description, I'd suggest trying out the option `std::vector<historyelement<Space>* > History;` first. I believe smart pointers are a C++11 feature. We've been trying to avoid C++11 features so as to not enforce additional technical requirements on people. I don't know enough about C++11 update to feel strongly about this but perhaps @eddelbuettel or @adamjohansen have thoughts on this?

I'd also recommend adding virtual functions to the base class if you haven't already. We're currently using inheritance for moveset so you might find it helpful to look at that part of the code. I've also found the derived classes chapter of Bjarne Stroustrup's book "The C++ Programming Language" helpful as a reference on this (and many other things).

I hope this is helpful. I'm a bit rusty on this but I think it'll help when I can see the code.

**Reply by eddelbuettel** [reply]
> We've been trying to avoid C++11 features so as to not enforce additional technical requirements on people

That actually no longer binds. R itself has a minimum of C++11 now, it will be C++14 next year.

The bigger issue I have is that I can't decide if fancier C++ classes are 'too fancy' in the sense that we can only ever return simpler things (in essence: vectors or matrices of ints or doubles (or chars) plus aggregations of these in lists or data.frames).   
Or would we want to persist the ancestral histories for other (later) session?  That would be a different issues, and set of headaches / questions.

So even bigger question:  what is that we currently want from our functions?

**Reply by adamjohansen** [reply]
First of all, draft is good... it makes more sense to discuss design before writing code than after lots of effort.

On pointers vs. smartpointers I'm slightly conflicted. I instinctively use pointers through habit and because they are simple, but the _right_ thing to do is almost certainly to use smart pointers with appropriate deep copying whenever this type of object is copied (which is likely to be when the overall sampler is duplicated in which case the right behaviour would be to arrive at a second sampler which shares the state of the original but which can subsequently evolve independently of it). I don't think we'd need any unusual persistence. 

Dirk's point is a good one and also relates to backward compatibility: if we modify the interface then existing code breaks and we should do that only if essential (which it may be). So, would it be possible, perhaps, to store things internally via smart pointers but dereference them automatically at the point at which users see them, returning simpler objects containing the information asked for? That would retain the keeping it simple from the user side philosophy as well as hopefully leaving the behaviour of existing interface functions as they are. (This might not make sense and I doubt that even if there are people who are using the existing codebase -- we need better documentation before it's likely that there will be many -- that very many are using the history storage features at present and we don't want to get too hung up on compatibility here.)

Another thought: I have a bad habit of resorting to polymorphism at the drop of a hat and almost certainly suggested it, but do we really need it here? As you suggested at the beginning one could simply add additional storage (which wouldn't be substantial) to the existing `historyelement` class and then avoid all the "fun" that comes with inheritance. Is there a good reason to go down the more complicated route? (I'm not saying that there isn't, simply asking the question.)

**Comment by ilyaZar** [comment]
Thanks! So I think we have two paths here (which I can push in separate branches of my fork, clever idea! thanks @LeahPrice):

1. As @adamjohansen suggested (hopefully I got you right) a 2-way version: simply change the existing `historyelement` class accordingly. We still have the option of not storing at all (`HistoryType::NONE`). This way we arrive at the two way version: no storage or storing particle system plus indices.

2. Going down the route with the 3-way version (NONE, RAM, AL) and all the associated pointer fun. This way, we will arrive at: no storage, storing particle system as before or storing particle system as before plus indices. Regarding this:
- @LeahPrice I always thought a virtual function is useful if a base class member function is overriden by a derived class. I guess this is very useful for the `moveset` class because the user wants to implement different SMC algorithms i.e. different proposal and or mcmc move schemes. In contrast, the derived `historyelementAL` does not require a change to the functionality of its base `historyelement` members; it simply adds on top: storage for indices and getters/setters for the indices. But it still may be useful here, I am not sure.
- @eddelbuettel As exposed to the user, I think, the functions pretty much do the same as before but additionally store the integer container for the ancestor indices and make them accessible (for later purposes e.g. implementing **_conditional_** SMC). We only change internal behavior in case there is merit to the 3-way version (i.e. the use case of storing only the particle system without indices via `HistoryType::RAM`).
- @adamjohansen Storing things internally via smart pointers but dereference at the correct places is something I was thinking about too: so currently the user interacts with 
```
///Returns the History of the particle system
const std::vector<historyelement<Space> > & GetHistory(void) const { return History; }
///Returns the number of evolution times stored in the history.
long GetHistoryLength(void) const {return History.size();}
///Returns the current particle set stored in the history.
population<Space> GetHistoryPopulation(long n) const {return History[n].GetValues();}
///Returns a reference to the particle set stored in the history.
const population<Space> & GetHistoryPopulationRefs(long n) const {return History[n].GetValues();}
// const arma::Col<unsigned int> GetHistoryPopulationAL(long n) const
// {return History[n].GetAindices();}
///Returns the history flags
historyflags GetHistoryFlags(long n) const {return History[n].GetFlags();}
/// Returns the Effective Sample Size of the specified particle generation.
double GetESS(long n) const {return History[n].GetESS();}
///Returns the history number of MCMC iterations performed during this iteration.
int GetHistorymcmcRepeats(long n) {return History[n].mcmcRepeats();}
```
And probably the output streams. The rest (sampler copy constructor and integration functions) is internal as far as I understood.

**_Questions:_**

1. At this point it's probably useful to know, if storing the whole particle system (without the indices i.e. `HistoryType::RAM`) would be a reasonable use case compared to storing particle system plus indices (`HistoryType::AL`) i.e. if the 3-way version has practical benefit. Probably storage cost is indeed negligible as @adamjohansen suggested, even for large particle systems? (I have no idea about this to be honest...) 

2. Depending on Q1, maybe stick for now to the 2-way version leaving the 3-way version for later? 

3. What would be the correct git/github way of maintaining my implementation for the 2-way version (which will probably be a pull request and then eventually be part of the master branch of the main repo) and my current implementation efforts of the 3-way version. The 3way version probably as a branch of my local repo and my remote fork waiting to be merged into the main repo or  get deleted later, something like that?

**Reply by adamjohansen** [reply]
My opinions for what they're worth:

1. Storage costs associated with keeping both indices and states will only really be significant in settings in which the particle "value" is a small quantity. Even there it's unlikely it would be more than 50% of the cost and this is a regime in which spatial costs probably aren't a major constraint. It would be possible to allow all three options anyone and simply not make use of the additional storage provide for the indices if it isn't asked for; I suggest it only because it might give a slight performance improvement if the indices aren't wanted but the simplicity afforded by storing the indices perhaps outweighs this. There are settings in which having access to approximations of all of the history of the particle system in terms of the collection of values and weights at each iteration is useful, and then indices aren't needed, but they aren't that numerous and it doesn't hurt much to have the index information as well.
2. Seems reasonable to me.
3. To my mind keeping it as a branch of your repository until such a time as its clear whether you want to merge it with the master branch or not seems a reasonable way to proceed.

In a sense having a bit of potentially redundant storage but simpler inner workings is probably a better solution in terms of the trade-off between functionality and development time. We could then, if we wanted to include the tree-based storage option for efficiency, consider making it work by having both a vector and a tree to store the history but leaving (at least) one of them empty; the cost of this relative to a more complex implementation seem at first glance to be negligible, although I could be missing something.

There may well be other perspectives, though.

**Comment by ilyaZar** [comment]
Allright, thanks for the answers! So in terms of code implementation:

1. I hope the commit message is useful and explains most of the things: https://github.com/ilyaZar/rcppsmc/commit/263dfb5b5d8094d730b75a1a0ad44879f45af033. Some additional remarks
- I thought a bit about the purposes of `GetApop()` and it might be useful, just in principle, to get access to the whole population classes already re-ordered according to the ancestral lines, because the population classes have the weights and particle values etc; or it might be completely unuseful... an alternative would be to write Getters for every possible quantity we want to have (like GetAweights to access weights of ancestral lines etc.). 
- To simply retrieve an ancestral line based on an index (e.g. when drawn via particle Gibbs) `GetALineSpace()` is sufficient which returns an `std::vector<Space>` of length of the current history saved. (this, plus a random index draw, could be used to overwrite the conditioning data field of a derived conditional SMC class).

2. To test the validity of the implementation under 1. there is an example that illustrates and tests the functionality of the new additions. On a cpp-level by printing ancestral lines, particle values, ancestor indices and re-mapped ancestor indices, and, on the R level, this implementation is run and, additionally, a plot of the ancestral lines is generated. It's in my package SVmodelRcppSMC, the corresponding commit is https://github.com/ilyaZar/SVmodelRcppSMC/commit/ab906a9663308e15897cd631c42d92d7ead08c31.

3. After installing SVmodelRcppSMC, an example illustrating the tracking of ancestral lines for the SV model is in the examples package SVmodelExamples: 09_example.R (https://github.com/ilyaZar/SVmodelExamples). It generates the ancestral line plot for simulated data from the SV model and one can play arround with the resampling frequency to visualize sample degeneration if resampling is performed frequently.

I am looking forward to all remarks and suggestion!

**Reply by adamjohansen** [reply]
Thanks! It looks good to me. I've had a look at the commit and made a few trivial comments, but I don't have a lot to add, really.

I haven't had a chance to look a 2. or 3. yet, but I'll try to do that this evening.

**Reply by LeahPrice** [reply]
Thanks @ilyaZar, this looks great.

I don't fully understand the purpose of GetApop yet, so I'd benefit from the additional documentation that Adam has suggested there.

I took a look through points 2-3 and the visualisation is very nice. It would be nice to get some examples of how to use this functionality into the RcppSMC package. We'll naturally be using this new functionality for some of the improvements we've got planned and I think we should add something like this visualisation too. I think this could all be done later.

It looks there might be a couple of errors in the `resample_freq` documentation.

**Reply by ilyaZar** [reply]
Thank you all! So I added all the suggestions from @adamjohansen , see here https://github.com/ilyaZar/rcppsmc/commit/9471487bc7632ec3ca241a7517d9a7d8bc8bf6d4

@LeahPrice absolutely agreed!
- I don't mind adding the example to RcppSMC now or later. Perhaps later, when it becomes clearer what is actually a reasonable addition to the overall library? What do the others think?
- added proper documentation to `resample_freq`!

I think `GetApop()` causes more headache than use so either:
1. We could discuss a bit what we can make out of it https://github.com/ilyaZar/rcppsmc/blob/9471487bc7632ec3ca241a7517d9a7d8bc8bf6d4/inst/include/sampler.h#L416-L443
2. Move it to my branch that contains the attempt to include historyelement as a derived class might be more reasonable, since we already have a "working" version?

So are there any other suggestions? Otherwise PR and move on to the conditional SMC implementation (or do we prefer the tree based storage version first)? 

I'd prefer to open a new discussion topic on cond. SMC (new feature, new topic; but we can continue to discuss the tree based ancestor storage version here, to compare it with the current implementation though).

**Comment by eddelbuettel** [comment]
But spell it `resampleFreq` ?  :grinning:   

No views on where to go with `GetApop()`.  Could also leave it in trunk and label it (in code comments) as preliminary or experimental and note any and all shortcomings.  I find that code relegated to branches eventually just disappears....

**Reply by ilyaZar** [reply]
Yes :+1: . Still part of the my other package, but when I merge this I'll change everything to camelCase.

**Comment by adamjohansen** [comment]
Personally, I don't have any problem with leaving `GetApop` in the trunk; the fact it took me a few minutes to work out what it did / was for doesn't mean its not useful or that we need to do anything different with it. I was just making a casual remark that when we come to write proper documentation (we really must do this one day...) that it might need a little bit of care to make it clear.

PR sounds good.

I'd suggest getting conditional SMC working first before worrying about tree-based implementations (which strike me as nice to have rather than core functionality).

**Comment by LeahPrice** [comment]
Thanks @ilyaZar! The only thing I see remaining is the edge case of `resample_freq = 1` in the documentation. I think this should be "if in [0,1)" and then "when it is greater than or equal to 1". In other words, the `resample_freq` is treated as an ESS threshold at 1 rather than as a proportion. Perhaps this very minor update could be done along with Dirk's suggested switch to `resampleFreq` when you're merging things.
